File: qemu-run.py
#!/usr/bin/env python3
"""Copyright (C) 2020 Lucie Cupcakes <lucie_linux [at] protonmail.com>
This file is part of qemu-run <https://gitlab.com/lucie_cupcakes/pybd>.
qemu-run is free software; you can redistribute it and/or modify it under
the terms of the GNU General Public License as published by the Free
Software Foundation; either version 3, or (at your option) any later
version.
qemu-run is distributed in the hope that it will be useful, but WITHOUT ANY
WARRANTY; without even the implied warranty of MERCHANTABILITY or
FITNESS FOR A PARTICULAR PURPOSE.  See the GNU General Public License
for more details.
You should have received a copy of the GNU General Public License
along with qemu-run; see the file LICENSE.  If not see <http://www.gnu.org/licenses/>."""

# Shared folders: Needs smbd and gawk
import os,sys,errno,socket,subprocess,time,base64,uuid
from enum import Enum

# Mini error handling lib:
class InfoMsgType(Enum):
    im_error=0
    im_warning=1
    im_info=2

class InfoMsg:
    def __init__(self,msg_txt,msg_type=InfoMsgType.im_error):
        self.msg_txt=msg_txt
        self.msg_type=msg_type

class ReturnCode:
    def __init__(self,ok=True,msg=None): #op_success=Boolean,infomsg=InfoMsg class.
        self.ok=ok
        self.msgs=[]
        if msg!=None:
            self.msgs.append(msg)
    def set_error(self,err_msg):
        self.ok=False
        self.msgs.append(err_msg)

# My replacement for configobj,fuck that shit.
def load_cfg_process(res,lines):
    for line in lines:
        if line.find('=')!=-1:
            sline=line.split('=', 1)
            key=sline[0].strip()
            if key.find('#')==-1:
                value=sline[1].strip()
                res[key]=value
    return res

def load_cfg_from_file(fpath,defaults=None): # defaults must be a dictionary..
    res={}
    if defaults is not None:
        res=defaults.copy()
    with open(fpath,'r') as fh:
        res=load_cfg_process(res,fh.readlines())
    return res

def load_cfg_from_args(arg_cfg_str,defaults=None,delimiter=','):
    res={}
    if defaults is not None:
        res=defaults.copy()
    res=load_cfg_process(res,arg_cfg_str.split(delimiter))
    return res

def get_disk_format(file_path):
    out=subprocess.check_output(['qemu-img','info',file_path],universal_newlines=True).split()
    prev_was_fmt=False
    result=''
    for s in out:
        if prev_was_fmt:
            result=s
            break
        else:
            prev_was_fmt=s=='format:' 
    return result

def spawn_daemon(func,args=None):
    # do the UNIX double-fork magic,see Stevens' 'Advanced
    # Programming in the UNIX Environment' for details (ISBN 0201563177)
    try:
        pid=os.fork()
        if pid > 0:
            return
    except OSError as e:
        print('fork #1 failed: %d (%s)' % (e.errno,e.strerror),file=sys.stderr)
        sys.exit(1)
    os.setsid()
    try:
        pid=os.fork()
        if pid > 0:
            sys.exit(0)
    except OSError as e:
        print('fork #2 failed: %d (%s)' % (e.errno,e.strerror),file=sys.stderr)
        sys.exit(1)
    if args!=None:
        func(args)
    else:
        func()
    os._exit(os.EX_OK)

def get_usable_port():
    port=0
    sock=socket.socket(socket.AF_INET,socket.SOCK_STREAM)
    sock.bind(('127.0.0.1',0))
    port=sock.getsockname()[1]
    sock.close()
    return port

def detect_if_cfg_is_in_args():
    res=False
    if len(sys.argv) >= 2:
        if sys.argv[1].find('--cfg=')!=-1:
            res=True
    return res
    
def check_if_file_exists_in_path(fname, resolve=False):
    env_path=os.environ['PATH'].split(':')
    for p in env_path:
        fpath='{}/{}'.format(p,fname)
        if os.path.isfile(fpath):
          return fpath if resolve else True  
    return '' if resolve else False

def program_find_vm_location():
    # return values=
    rc=ReturnCode()
    vm_name=''
    vm_dir=''
    # Use environment variable to find where the VM is
    if len(sys.argv)==1: # No args.
        rc.set_error(InfoMsg('Warning: No arguments,assuming the VM is in CWD.',InfoMsgType.im_warning))
        vm_dir=os.getcwd()
    else:
        if detect_if_cfg_is_in_args()==True:
            vm_dir=os.getcwd()
            vm_name='QEMU-VM'
        else:
            # Normal lookup using ENV var
            try:
                vm_dir_env=os.environ['QEMURUN_VM_PATH'].split(':')
            except:
                rc.set_error(InfoMsg('Cannot find environment variable QEMURUN_VM_PATH.\nPython error \#%d (%s)' % (e.errno,e.strerror)))
            vm_name=sys.argv[1]
            for p in vm_dir_env:
                vm_dir='{}/{}'.format(p,vm_name)
                if os.path.exists(vm_dir):
                    break
            if not os.path.exists(vm_dir):
                rc.set_error(InfoMsg('Cannot find VM: {},Check your VM_PATH env. variable ?.'.format(vm_dir)))
    return rc,vm_name,vm_dir

def program_get_cfg_values(vm_dir):
    rc=ReturnCode()
    cfg={}
    cfg['*']=''
    cfg['sys']='x86_64'
    cfg['uefi']='no'
    cfg['cpu']='host'
    cfg['cores']=subprocess.check_output(['nproc']).decode(sys.stdout.encoding).strip()
    cfg['mem']='2G'
    cfg['acc']='yes'
    cfg['vga']='virtio'
    cfg['snd']='hda'
    cfg['boot']='c'
    cfg['fwd_ports']=''
    cfg['hdd_virtio']='yes'
    cfg['shared']='shared'
    cfg['net']='virtio-net-pci'
    cfg['rng_dev']='yes'
    cfg['host_video_acc']='no'
    cfg['localtime']='no'
    cfg['headless']='no'
    cfg['vnc_pwd']=''
    cfg['monitor_port']=5510
    cfg['floppy']='{}/floppy'.format(vm_dir) if os.path.isfile('{}/floppy'.format(vm_dir)) else 'No'
    cfg['cdrom']='{}/cdrom'.format(vm_dir) if os.path.isfile('{}/cdrom'.format(vm_dir)) else 'No'
    cfg['disk']='{}/disk'.format(vm_dir) if os.path.isfile('{}/disk'.format(vm_dir)) else 'No'
    if os.path.exists('{}/{}'.format(vm_dir,cfg['shared'])):
        cfg['shared']='{}/{}'.format(vm_dir,cfg['shared'])
    vm_cfg_file_path='{}/config'.format(vm_dir)
    if os.path.isfile(vm_cfg_file_path):
        cfg=load_cfg_from_file(vm_cfg_file_path,cfg)
    else:
        if detect_if_cfg_is_in_args()==True:
            arg_cfg_str=sys.argv[1].split('--cfg=')[1]
            rc.set_error(InfoMsg('Info: Using configuration from arguments.',InfoMsgType.im_info))
            if arg_cfg_str.strip()=='':
                rc.set_error(InfoMsg('Error: argument cfg is empty.'))
                return rc,None
            cfg=load_cfg_from_args(arg_cfg_str,cfg)
        else:
            rc.set_error(InfoMsg('Cannot find config file.'))
    return rc,cfg

def program_build_cmd_line(cfg,vm_name,vm_dir):
    rc=ReturnCode()
    drive_index=0
    telnet_port=0
    sf_str=''
    fwd_ports_str=''
    qemu_cmd=[]
    qemu_bin='qemu-system-{}'.format(cfg['sys'].lower())
    vm_is_x86=cfg['sys'].lower()=='x86_64' or cfg['sys'].lower()=='i386'
    if check_if_file_exists_in_path(qemu_bin): 
        qemu_cmd.append(qemu_bin)
    else:
        rc.set_error(InfoMsg('Invalid sys value.'))
    if cfg['acc'].lower()=='yes' and vm_is_x86:
        qemu_cmd.append('--enable-kvm')
    if vm_name!='':
        qemu_cmd+=['-name',vm_name]
    if cfg['uefi'].lower()=='yes':
        qemu_cmd+=['-L','/usr/share/qemu','-bios','OVMF.fd']
    if not (cfg['cpu'].lower()=='host' and vm_is_x86==False):
        qemu_cmd+=['-cpu',cfg['cpu']]
    qemu_cmd+=['-smp',cfg['cores'],
                '-m',cfg['mem'],
                '-boot','order=' + cfg['boot'],
                '-usb','-device','usb-tablet']
    if cfg['snd'].lower()!='no':
        qemu_cmd+=['-soundhw',cfg['snd']]
    qemu_cmd+=['-vga',cfg['vga']]
    if cfg['headless'].lower()=='yes':
        telnet_port=get_usable_port()
        qemu_cmd+=['-monitor','telnet:127.0.0.1:{},server,nowait'.format(telnet_port)]
        if cfg['vnc_pwd']!='':
            qemu_cmd+=['-vnc','127.0.0.1:0,password']
        else:
            qemu_cmd+=['-vnc','127.0.0.1:0']
        qemu_cmd+=['-display','none']
    else:
        if cfg['host_video_acc'].lower()=='yes':
            qemu_cmd+=['-display','gtk,gl=on']
        else:
            qemu_cmd+=['-display','gtk,gl=off']
    if cfg['rng_dev'].lower()=='yes':
        qemu_cmd+=['-object','rng-random,id=rng0,filename=/dev/random','-device','virtio-rng-pci,rng=rng0']
    if os.path.exists(cfg['shared']):
        sf_str=',smb=' + cfg['shared']
    if cfg['fwd_ports']!='':
        ports_split=cfg['fwd_ports'].split(',')
        for pair_str in ports_split:
            if pair_str.find(':')!=-1: # If have FordwardPorts=<HostPort>:<GuestPort>
                pair=pair_str.split(':')
                fwd_ports_str+=',hostfwd=tcp::{}-:{},hostfwd=udp::{}-:{}'.format(pair[0],pair[1],pair[0],pair[1]);
            else:   # Else use the same port for Host and Guest.
                fwd_ports_str+=',hostfwd=tcp::{}-:{},hostfwd=udp::{}-:{}'.format(pair_str,pair_str,pair_str,pair_str);
    qemu_cmd+=['-nic','user,model={}{}{}'.format(cfg['net'],sf_str,fwd_ports_str)]
    if os.path.isfile(cfg['floppy']):
        qemu_cmd+=['-drive','index={},file={},if=floppy,format=raw'.format(str(drive_index),cfg['floppy'])]
        drive_index+=1
    if os.path.isfile(cfg['cdrom']):
        qemu_cmd+=['-drive','index={},file={},media=cdrom'.format(str(drive_index),cfg['cdrom'])]
        drive_index+=1
    if os.path.isfile(cfg['disk']):
        hdd_fmt=get_disk_format(cfg['disk'])
        hdd_virtio=''
        if cfg['hdd_virtio'].lower()=='yes':
            hdd_virtio=',if=virtio'
        qemu_cmd+=['-drive','index={},file={},format={}{}'.format(str(drive_index),cfg['disk'],hdd_fmt,hdd_virtio)]
        drive_index+=1
    if cfg['localtime']=='Yes':
        qemu_cmd+=['-rtc','base=localtime']
    if cfg['*']!='':
        qemu_cmd+=cfg['*'].split(' ')
    return rc,qemu_cmd,telnet_port

def program_handle_rc(rc):
    if rc.ok==False:
        for msg in rc.msgs:
            if msg.msg_type==InfoMsgType.im_error:
                print(msg.msg_txt)
                exit()
            elif msg.msg_type==InfoMsgType.im_warning or msg.msg_type==InfoMsgType.im_info:
                print(msg.msg_txt)

def execute_bash_code(contents):
    fpath="/tmp/qemurun_{}.sh".format(str(uuid.uuid4()))
    fh=open(fpath,"w+")
    fh.write(contents)
    fh.close()
    env_cpy=os.environ.copy()
    fnull=open(os.devnull,'w')
    rc=subprocess.call(['bash',fpath],env=env_cpy,stdout=fnull,stderr=fnull)
    os.remove(fpath)
    return rc

def program_subprocess_qemu(args):
    sp=subprocess.Popen(args['qemu_cmd'],env=args['qemu_env'],cwd=args['vm_dir'])
    print('QEMU Running at PID: {}'.format(str(sp.pid)))
    if args['telnet_port']!=0:
        print('Telnet monitor port: {}'.format(str(args['telnet_port'])))
    return sp

def program_subprocess_fix_smb():
    time.sleep(5)
    script="""#!/bin/bash
        smb_dir=$(ls /tmp | grep qemu-smb | head -n1)
        if [ "$smb_dir"!="" ]; then
            smb_dir="/tmp/$smb_dir" 
            echo "[global]
                allow insecure wide links=yes
                [qemu]
                follow symlinks=yes
                wide links=yes
                acl allow execute always=yes" >> "$smb_dir/smb.conf"
            #smbcontrol --configfile=$conf $pid reload-config
            exit 0
        fi"""
    return execute_bash_code(script)
    #return True

def program_subprocess_change_vnc_pwd(args):
    time.sleep(5)
    script="""#!/bin/bash
    vnc_pwd='@vnc_pwd@'
    printf "change vnc password\n%s\nexit\n" $vnc_pwd | nc 127.0.0.1 @telnet_port@"""
    script=script.replace("@vnc_pwd@",args['vnc_pwd'])
    script=script.replace("@telnet_port@",str(args['telnet_port']))
    return execute_bash_code(script)

def program_main():
    print('qemu-run. Forever beta software. Use on production on your own risk!\n')
    print('This software is Free software - released under the GPLv3 License.')
    print('Read the LICENSE file. Or go visit https://www.gnu.org/licenses/gpl-3.0.html\n')
    rc,vm_name,vm_dir=program_find_vm_location()
    program_handle_rc(rc)
    rc,cfg=program_get_cfg_values(vm_dir)
    program_handle_rc(rc)
    telnet_port=0
    rc,qemu_cmd,telnet_port=program_build_cmd_line(cfg,vm_name,vm_dir)
    program_handle_rc(rc)
    qemu_env=os.environ.copy()
    qemu_env['SDL_VIDEO_X11_DGAMOUSE']='0'
    print('Command line arguments:')
    print(*qemu_cmd)
    if os.path.exists(cfg['shared']):
        spawn_daemon(program_subprocess_fix_smb)
    if cfg['vnc_pwd']!='':
        args={}
        args['vnc_pwd']=cfg['vnc_pwd']
        args['telnet_port']=telnet_port
        spawn_daemon(program_subprocess_change_vnc_pwd,args)
    args={}
    args['qemu_cmd']=qemu_cmd
    args['qemu_env']=qemu_env
    args['vm_dir']=vm_dir
    args['telnet_port']=telnet_port
    spawn_daemon(program_subprocess_qemu, args)

if __name__=='__main__':
    program_main()

File: qemu-run.c
#include <stdlib.h>
#include <stdio.h>
#include <stdarg.h>
#include <stdbool.h>
#include <string.h>
#include <ctype.h>

#if defined(__unix__) || defined(__unix) || defined(unix)
#define __NIX__
#define PSEP ":"
#define DSEP "/"
#include <unistd.h>
#ifdef __linux__
#include <linux/limits.h>
#else
#include <limits.h>
#endif // __linux__
#else // !__NIX__
#if defined(WIN32) || defined(_WIN32) || defined(__WIN32__) || defined(__NT__)
#define __WINDOWS__
#define PSEP ";"
#define DSEP "\\"
#include <io.h>
#include <limits.h>
#endif // __WINDOWS__
#endif // __NIX__
#include <sys/stat.h>
#include <sys/types.h>

#ifdef DEBUG
#ifdef __GNUC__
#define dprint() printf("%s (%d) %s\n",__FILE__,__LINE__,__FUNCTION__);
#else
#define dprint() printf("%s (%d) %s\n",__FILE__,__LINE__);
#endif
#else // !DEBUG
#define dprint()
#endif // DEBUG

#define BUFF_AVG 128
#define BUFF_MAX BUFF_AVG*32

#ifndef stricmp //GCC is weird sometimes it doesn't includes this..
#include <strings.h>
#define stricmp(x,y) strcasecmp(x,y)
#endif

#if defined(_SVID_SOURCE)||defined(_BSD_SOURCE)||_XOPEN_SOURCE >= 500||defined(_XOPEN_SOURCE)&&defined(_XOPEN_SOURCE_EXTENDED)
#else
char *strdup(const char *s) {
	size_t n=strlen(s)+1;
	char *p=malloc(n);
	return p ? memcpy(p,s,n) : p;
}
#endif

#include "config.h"

#define mzero_ca(a) memset(a, 0, sizeof(a));

char *strcatx(char *str, ...) {
	va_list ap;
	int dl,sl;
	char *arg;
	for(dl=0;str[dl];dl++);
	va_start(ap,str);
	while ((arg = va_arg(ap,char*))) {
		for(sl=0;arg[sl];sl++) { str[dl++]=arg[sl]; }
	}
	va_end(ap);
	str[dl]=0;
	return str;
}

char *bcitoa(char *sequence,int base,int num,char *out_buff) {
	char *dflt="0123456789ABCDEF";
	char *str,*vals;
	int max=1,len=1,idx=0;
	vals=(sequence==NULL)?dflt:sequence;
	while((max*base)<num) {max*=base;len++;}
	if(out_buff) {
	   str=out_buff;
	} else {
		str=(char *)calloc(1,len+1);
	}
	while(len--) {
		str[idx++]=vals[(num/max)];
		num-=(num/max)*max;
		max/=base;
   }
   return str;
}

int bcatoi(char *sequence,int base,char *str) {
   char *dflt="0123456789ABCDEF";
   char *val;
   int num=0,idx;
   val=(sequence==NULL)?dflt:sequence;
   while(*str) {
      for(idx=0;idx<base && *str!=val[idx];idx++);
      if(idx==base) {num=0;break;}
      num=(num*base)+idx;
      str++;
   }
   return num;
}

enum {ERR_UNKOWN,ERR_ARGS,ERR_ENV,ERR_CHDIR_VM_DIR,ERR_OPEN_CONFIG,ERR_FIND_CONFIG,ERR_SYS,ERR_NETCONF_IP,ERR_SHAREDF,ERR_EXEC,ERR_ENDLIST};
void fatal(unsigned int errcode) {
	char *errs[]={
		"Unkown error.",// ERR_UNKOWN
		"Invalid argument count. Did you specified the VM Name?",
		"Cannot find VM. Please check your QEMURUN_VM_PATH env. variable",
		"Cannot access VM folder. Maybe check its permissions",
		"Cannot find VM config file. Is it created?",
		"Cannot open config file. Maybe check file permissions",
		"Config file has an invalid value for sys",
		"Invalid configuration: VM has enabled network, but has IPv6 and IPv4 disabled. Please enable at least one",
		"Invalid configuration: VM has disabled network, and specified a shared folder. Enable network or disable the shared folder.",
		"There was an error trying to execute qemu. Is it installed?"
	};
	dprint();
	printf("There was an error in the program:\n\t%s.\n",errs[errcode<ERR_ENDLIST?errcode:0]);
	exit(1);
}

int sym_hash_generate(char *str) {
	int pos=0,hash=0xCAFEBABE;
	while(str[pos]) {
		hash=~(((((hash&0xFF)^str[pos])<<5)|
			(((hash>>26)&0x1f)^(pos&31)))|
			(hash<<18));
		pos++;
	}
	hash&=0xFFFFFFFF;
	return hash;
}

bool sym_put_kv(char *key,char *val) {
	int cnt=0,ret=0,hash=sym_hash_generate(key);
	for(;cnt<KEY_ENDLIST;cnt++) {
		if(cfg[cnt].hash==hash) { ret=1;cfg[cnt].val=strdup(val); }
	}
	return ret;
}

enum {FT_TYPE,FT_UNKNOWN=0,FT_PATH,FT_FILE};
int filetype(const char *fpath,int type) {
	int ret=0; struct stat sb;
	if((ret=(access (fpath,0)==0))) {
		stat(fpath,&sb);
		ret=S_ISREG(sb.st_mode)>0?FT_FILE:(S_ISDIR(sb.st_mode)?FT_PATH:FT_UNKNOWN);
	}
	return ret=(type)?ret==type:ret;
}

char* cstr_remove_quotes(char* c) {
	dprint();
	size_t len=strlen(c);
	if(c[len-1]=='\"'&&c[0]=='\"') {
		c[len-1]='\0';
		return c+1;
	}else{ return c; }
}

#ifdef __WINDOWS__ // For now I only need this function on Windows.
bool get_binary_full_path(char *bin_fname,char *out_bin_fpath,char *out_dir) {
	dprint();
	bool found = 0;
	char fp_b[PATH_MAX]={0}, dir_pb[PATH_MAX]={0}, *env;
	if (! (env=getenv("PATH"))){ fatal(ERR_EXEC); }
	char *dir_p = strtok(env, PSEP);
	while (dir_p && !found) {
		strcpy(dir_pb, dir_p);
		char* dir_pq = cstr_remove_quotes(dir_pb);
		strcatx(fp_b, dir_pq, DSEP, bin_fname, NULL); found = filetype(fp_b, FT_FILE);
		if (!found) { mzero_ca(fp_b); strcatx(fp_b, dir_pq, DSEP, bin_fname, ".exe", NULL); found = filetype(fp_b, FT_FILE); }
		if (!found) { mzero_ca(fp_b); strcatx(fp_b, dir_pq, DSEP, bin_fname, ".bat", NULL); found = filetype(fp_b, FT_FILE); }
		if (!found) { mzero_ca(fp_b); strcatx(fp_b, dir_pq, DSEP, bin_fname, ".com", NULL); found = filetype(fp_b, FT_FILE); }
		if (found && out_dir) { strcpy(out_dir, dir_pq); }
		dir_p = strtok(NULL, PSEP);
	}
	if(found&&out_bin_fpath) { strcpy(out_bin_fpath,fp_b); }
	return found;
}
#endif

void program_load_config(const char *fpath) {
	dprint();
	FILE *fptr=fopen(fpath,"r");
#ifdef DEBUG
	printf("fpath=%s\n",fpath);
#endif
	if(!fptr) { fatal(ERR_OPEN_CONFIG); }
	char line[BUFF_AVG*2]={0},key[BUFF_AVG]={0},val[BUFF_AVG]={0};
	while(fgets(line,BUFF_AVG*2,fptr)) {
		if(!strchr(line,'=') || line[0]=='#') { continue; }
		size_t len=strlen(line);
		if(len&&line[len-2]=='\r') { len--; line[len-1]='\0'; } 
		if(len&&line[len-1]=='\n') { len--; line[len]='\0'; }
		if(len<3) { continue; }
		char* slice=strtok(line,"=");
		for(int i=0; slice&&i<2; i++) {
			if(i==0) { strcpy(key,slice); }
			if(i==1) { strcpy(val,slice); break; }
			slice=strtok(NULL,"=");
		}
		sym_put_kv(key,val);
	}
	fclose(fptr);
}

void program_set_default_cfg_values() {
	dprint();
#ifdef __WINDOWS__
	sym_put_kv("acc","no");
	sym_put_kv("cpu","max");
	sym_put_kv("rng_dev","no");
#endif
	/* Because now qemu-run chdirs into vm_dir,
	it's not needed to append vm_dir to the filename. */
	if(filetype("shared",FT_PATH))			{ sym_put_kv("shared","shared"); }
	if(filetype("floppy",FT_FILE))			{ sym_put_kv("floppy","floppy"); }
	if(filetype("floppy.img",FT_FILE))		{ sym_put_kv("floppy","floppy.img"); }
	if(filetype("cdrom",FT_FILE))			{ sym_put_kv("cdrom","cdrom"); }
	if(filetype("cdrom.iso",FT_FILE))		{ sym_put_kv("cdrom","cdrom.iso"); }
	if(filetype("disk",FT_FILE))			{ sym_put_kv("disk","disk"); }
	if(filetype("disk.qcow2",FT_FILE))		{ sym_put_kv("disk","disk.qcow2"); }
	if(filetype("disk.raw",FT_FILE))		{ sym_put_kv("disk","disk.raw"); }
	if(filetype("disk.img",FT_FILE))		{ sym_put_kv("disk","disk.img"); }
	if(filetype("disk.vmdk",FT_FILE))		{ sym_put_kv("disk","disk.vmdk"); }
	if(filetype("disk.vdi",FT_FILE))		{ sym_put_kv("disk","disk.vdi"); }
	if(filetype("disk.vpc",FT_FILE))		{ sym_put_kv("disk","disk.vpc"); }
	if(filetype("disk.vhdx",FT_FILE))		{ sym_put_kv("disk","disk.vhdx"); }
}

void program_build_cmd_line(char *vm_name, char *out_cmd) {
	int drive_index = 0; // telnet_port = 55555; // @TODO: Get usable TCP port
	char drive_str[4]={0};
	dprint();
#ifdef __WINDOWS__
	char qemu_binary_file[64] = {0};
#else
	bool vm_has_rngdev=stricmp(cfg[KEY_RNG_DEV].val,"yes")==0;
#endif
	bool vm_has_name = (strcmp(vm_name, "") != 0 );
	bool vm_has_acc_enabled = stricmp(cfg[KEY_ACC].val, "yes")==0;
	bool vm_has_vncpwd = (strcmp(cfg[KEY_VNC_PWD].val, "") != 0);
	bool vm_has_audio = stricmp(cfg[KEY_SND].val, "no")!=0;
	bool vm_has_videoacc = stricmp(cfg[KEY_HOST_VIDEO_ACC].val, "yes")==0;
	bool vm_is_headless = stricmp(cfg[KEY_HEADLESS].val, "yes")==0;
	bool vm_clock_is_localtime = stricmp(cfg[KEY_LOCALTIME].val, "yes")==0;
	bool vm_has_sharedf = (strcmp(cfg[KEY_SHARED].val, "") != 0 && filetype(cfg[KEY_SHARED].val, FT_PATH));
	bool vm_has_hddvirtio = stricmp(cfg[KEY_HDD_VIRTIO].val, "yes")==0;
	bool vm_has_network = (strcmp(cfg[KEY_NET].val, "no") != 0 );
	bool vm_has_fwd_ports = strcmp(cfg[KEY_FWD_PORTS].val, "no") != 0;
	bool vm_has_ipv4 = stricmp(cfg[KEY_IPV4].val, "yes")==0;
	bool vm_has_ipv6 = stricmp(cfg[KEY_IPV6].val, "yes")==0;
	vm_has_sharedf = vm_has_sharedf ? filetype(cfg[KEY_SHARED].val,FT_PATH) : 0;

	if (strcmp(cfg[KEY_SYS].val, "x32") == 0) {
		strcpy(out_cmd, "qemu-system-i386");
#ifdef __WINDOWS__
		strcpy(qemu_binary_file, "qemu-system-i386");
#endif
	} else if (strcmp(cfg[KEY_SYS].val, "x64") == 0) {
		strcpy(out_cmd, "qemu-system-x86_64");
#ifdef __WINDOWS__
		strcpy(qemu_binary_file, "qemu-system-x86_64");
#endif
	} else { fatal(ERR_SYS); }
	
	if(vm_has_acc_enabled) { strcat(out_cmd, " --enable-kvm"); }
	if(vm_has_name) { strcatx(out_cmd, " -name ",  vm_name, NULL); }

	strcatx(out_cmd,
		" -cpu ", cfg[KEY_CPU].val,
		" -smp ", cfg[KEY_CORES].val,
		" -m ", cfg[KEY_MEM].val,
		" -boot order=", cfg[KEY_BOOT].val,
		" -usb -device usb-tablet -vga ", cfg[KEY_VGA].val,
		NULL);

	if(vm_has_audio) { strcatx(out_cmd, " -soundhw ", cfg[KEY_SND].val, NULL); }

	if (vm_is_headless) {
		strcatx(out_cmd, " -display none -monitor telnet:127.0.0.1:55555,server,nowait -vnc 127.0.0.1:0", vm_has_vncpwd ? ",password" : "", NULL);
	} else {
		strcatx(out_cmd, vm_has_videoacc ? " -display gtk,gl=on" : " -display gtk,gl=off", NULL);
	}

	if (vm_has_network && !vm_has_ipv6 && !vm_has_ipv4) { fatal(ERR_NETCONF_IP); }
	if (!vm_has_network && vm_has_sharedf) { fatal(ERR_SHAREDF); }
	if (vm_has_network) {
		strcatx(out_cmd,
				" -nic user,model=", cfg[KEY_NET].val,
				vm_has_ipv4 ? ",ipv4=on" : ",ipv4=off",
				vm_has_ipv6 ? ",ipv6=on" : ",ipv6=off",
				NULL);
	}
	if (vm_has_network && vm_has_sharedf) { strcatx(out_cmd, ",smb=", cfg[KEY_SHARED].val, NULL); }
	if (vm_has_network && vm_has_fwd_ports) {
		char fwd_port_a[16], fwd_port_b[16];
		char* cfg_v_c = strdup(cfg[KEY_FWD_PORTS].val);
		if (strchr(cfg_v_c, ':') != NULL) { // If have fwd_ports=<HostPort>:<GuestPort>
			char *fwd_ports_tk = strtok(cfg_v_c, ":");
			for (int i = 0; fwd_ports_tk && i<2; i++) {
				strcpy(i == 0 ? fwd_port_a : fwd_port_b, fwd_ports_tk);
				fwd_ports_tk = strtok(NULL, ":");
			}
			strcatx(out_cmd,",hostfwd=tcp::",fwd_port_a,"-:",fwd_port_b,",hostfwd=udp::",fwd_port_a, "-:",fwd_port_b, NULL);
		} else { // Else use the same port for Host and Guest.
			strcatx(out_cmd,",hostfwd=tcp::",cfg_v_c,"-:",cfg_v_c,",hostfwd=udp::",cfg_v_c, "-:",cfg_v_c, NULL);
		}
	}

	if (filetype((const char*) cfg[KEY_FLOPPY].val,FT_FILE)) {
		bcitoa(NULL, 10, drive_index, drive_str);
		strcatx(out_cmd, " -drive index=", drive_str, ",file=", cfg[KEY_FLOPPY].val, ",format=raw", NULL);
		drive_index++;
	}

	if (filetype(cfg[KEY_CDROM].val,FT_FILE)) {
		bcitoa(NULL, 10, drive_index, drive_str);
		strcatx(out_cmd, " -drive index=", drive_str, ",file=", cfg[KEY_CDROM].val, ",media=cdrom", NULL);
		drive_index++;
	}

	if (filetype(cfg[KEY_DISK].val,FT_FILE)) {
		bcitoa(NULL, 10, drive_index, drive_str);
		strcatx(out_cmd, " -drive index=", drive_str, ",file=", cfg[KEY_DISK].val, vm_has_hddvirtio ? ",if=virtio" : "", NULL);
		drive_index++;
	}

#ifdef __WINDOWS__
	/* QEMU on Windows needs,for some reason,to have an additional argument with the program path on it,
	* For example if you have it on: "C:\Program Files\Qemu",You have to run it like this: qemu-system-i386.exe -L "C:\Program Files\Qemu"
	* Otherwise it wont find the BIOS file.. */
	char qemu_binary_full_path[BUFF_AVG]={0};
	char *qemu_binary_full_path_p = &qemu_binary_full_path[0];
	if (! get_binary_full_path(qemu_binary_file, NULL, qemu_binary_full_path_p)) { fatal(ERR_EXEC); }
	strcatx(out_cmd, " -L \"", qemu_binary_full_path_p, "\"", NULL);
#else
	if (vm_has_rngdev) { strcat(out_cmd, " -object rng-random,id=rng0,filename=/dev/random -device virtio-rng-pci,rng=rng0"); }
#endif
	if (vm_clock_is_localtime) { strcat(out_cmd, " -rtc base=localtime"); }
}

void program_find_vm_and_chdir(int argc,char **argv,char *out_vm_name,char *out_vm_cfg_file) {
	char vm_dir[PATH_MAX+1],env_dir[PATH_MAX]={0},*env,*env_dir_p;
	bool vm_dir_exists=0,cfg_file_exists=0;
	dprint();
	if (argc < 2) { fatal(ERR_ARGS); }
	strcpy(out_vm_name, argv[1]);
	if (!(env = getenv("QEMURUN_VM_PATH"))) { fatal(ERR_ENV); }
	if (strcmp(env, "") == 0) { fatal(ERR_ENV); }
	env_dir_p = strtok(env, PSEP);
	while (env_dir_p && !vm_dir_exists) {
		strcpy(env_dir, env_dir_p);
		char* env_dir_q = cstr_remove_quotes(env_dir);
		mzero_ca(vm_dir);
		strcatx(vm_dir, env_dir_q, DSEP, out_vm_name, NULL);
		vm_dir_exists = filetype(vm_dir,FT_PATH);
		env_dir_p = strtok(NULL, PSEP);
	}
	if(! vm_dir_exists) { fatal(ERR_ENV); }
	if(chdir(vm_dir) != 0) { fatal(ERR_CHDIR_VM_DIR); }
	strcpy(out_vm_cfg_file,"config"); cfg_file_exists=filetype(out_vm_cfg_file,FT_FILE);
	if(! cfg_file_exists) { strcpy(out_vm_cfg_file,"config.ini"); cfg_file_exists=filetype(out_vm_cfg_file,FT_FILE); }
	if(! cfg_file_exists) { fatal(ERR_FIND_CONFIG); }
}

int main(int argc,char **argv) {
	char cmd[BUFF_MAX],vm_name[BUFF_AVG],vm_cfg_file[BUFF_AVG];
	dprint();
	program_find_vm_and_chdir(argc,argv,vm_name,vm_cfg_file);
	program_set_default_cfg_values();
	program_load_config(vm_cfg_file);
#ifdef DEBUG
	puts("Hash table:"); for(int i=0;i<KEY_ENDLIST;i++) printf("%d 0x%8.8X='%s'\n",i,cfg[i].hash,cfg[i].val);
#endif
	program_build_cmd_line(vm_name,cmd);
	puts("QEMU Command line arguments:");
	puts(cmd);
	return system(cmd); // Just Pass Qemu's Err code on
}
